### PR TITLE
Function to download mp3 file locally

### PR DESCRIPTION
**Pull request by sandlerben**
See #22

### CODE REVIEW

**Comment by sandlerben** [comment]
Made this so code review would be easier

File: transcription/transcription.go [hunk 1]
```diff
@@ -29,3 +32,32 @@ func msgHeaders(from string, to []string, subject string) string {
...
 	return strings.Join(msgHeaders, "\r\n")
 }
+
+// DownloadFileFromURL locally downloads an audio file stored at url.
+func DownloadFileFromURL(url string) error {
+	// Taken from https://github.com/thbar/golang-playground/blob/master/download-files.go
+	tokens := strings.Split(url, "/")
+	fileName := tokens[len(tokens)-1]
+
+	// Create the file
```
Comment (sandlerben) [inline transcription/transcription.go]: Make sure to remove these unnecessary comments. Opt for self-documenting code instead!

File: transcription/transcription.go [hunk 2]
```diff
@@ -29,3 +32,32 @@ func msgHeaders(from string, to []string, subject string) string {
 	msgHeaders := []string{fromHeader, toHeader, subjectHeader}
 	return strings.Join(msgHeaders, "\r\n")
 }
+
+// DownloadFileFromURL locally downloads an audio file stored at url.
+func DownloadFileFromURL(url string) error {
+	// Taken from https://github.com/thbar/golang-playground/blob/master/download-files.go
+	tokens := strings.Split(url, "/")
```
Comment (sandlerben) [inline transcription/transcription.go]: What is this hoohaa with tokens?

**Comment by yoninachmany** [comment]
Thanks! Should I create mock tests? Not really sure why I would mock `http.Get` or `io.Copy`

File: transcription/transcription.go [hunk 3]
```diff
@@ -29,3 +32,32 @@ func msgHeaders(from string, to []string, subject string) string {
...
+	fileName := tokens[len(tokens)-1]
```
Comment (sandlerben) [inline transcription/transcription.go]: This logic seems confusing enough that I would put it inside a `fileNameFromUrl` function.

File: transcription/transcription.go [hunk 4]
```diff
@@ -29,3 +32,32 @@ func msgHeaders(from string, to []string, subject string) string {
...
+
+	// Create the file
+	output, err := os.Create(fileName)
+	if err != nil {
+		return err
+	}
+	defer output.Close()
```
Comment (sandlerben) [inline transcription/transcription.go:43]: :')

**Comment by sandlerben** [comment]
If you don't mock those functions the code will make an _actual_ get request (meaning our tests would require internet) and copy an actual file.

**Comment by sandlerben** [comment]
Plus that would introduce non-determinism into our tests.

File: transcription/transcription.go (hunk 1 above)
Comment (yoninachmany) [inline transcription/transcription.go]: So @anoshaminai commented on this, asking why it was unnecessary if every other section of code had overarching comment about its purpose.

She recommended changing comment to // Get file contents. I thought I pushed this, I guess not -- would that comment still be unnecessary?

File: transcription/transcription.go (hunk 3 above)
Comment (yoninachmany) [inline transcription/transcription.go]: rather than just a comment above these two lines about what is happening?

**Comment by yoninachmany** [comment]
Right now, determinism in mocking comes from mocking `smtp` functions from Go library, which ideally should be working, but I guess I understand that we should test them.

I guess my question is: shouldn't we continuously test the real thing - sending an actual email or downloading an actual url?

File: transcription/transcription.go (hunk 1 above)
Comment (sandlerben) [inline transcription/transcription.go]: > So @anoshaminai https://github.com/anoshaminai commented on this,
> asking why it was unnecessary if every other section of code had
> overarching comment about its purpose.

I think this code is pretty understandable without comments.

On Friday, March 4, 2016, Yoni Nachmany notifications@github.com wrote:

> In transcription/transcription.go
> https://github.com/hack4impact/audio-transcription-service/pull/24#discussion_r55069975
> :
> 
> > @@ -29,3 +32,32 @@ func msgHeaders(from string, to []string, subject string) string {
> >     msgHeaders := []string{fromHeader, toHeader, subjectHeader}
> >     return strings.Join(msgHeaders, "\r\n")
> >  }
> > +
> > +// DownloadFileFromURL locally downloads an audio file stored at url.
> > +func DownloadFileFromURL(url string) error {
> > -   // Taken from https://github.com/thbar/golang-playground/blob/master/download-files.go
> > -   tokens := strings.Split(url, "/")
> > -   fileName := tokens[len(tokens)-1]
> >   +
> > -   // Create the file
> 
> So @anoshaminai https://github.com/anoshaminai commented on this,
> asking why it was unnecessary if every other section of code had
> overarching comment about its purpose.
> 
> She recommended changing comment to // Get file contents. I thought I
> pushed this, I guess not -- would that comment still be unneccessary?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/hack4impact/audio-transcription-service/pull/24/files#r55069975
> .

## 

Benjamin Sandler
University of Pennsylvania | Class of 2018
Candidate for BS in Computer Science & BA in Cognitive Science
954.478.6689 | ben1sandler@gmail.com

**Comment by sandlerben** [comment]
> Thanks! Should I create mock tests? Not really sure why I would mock http.Get or io.Copy

You would mock these things so that our application wouldn't copy to a file
or make an http get request during test. I'm actually okay with not mocking
io.Copy as long as the file is deleted as part of the clean up.

On Friday, March 4, 2016, Yoni Nachmany notifications@github.com wrote:

> Thanks! Should I create mock tests? Not really sure why I would mock
> http.Get or io.Copy
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/hack4impact/audio-transcription-service/pull/24#issuecomment-192363779
> .

## 

Benjamin Sandler
University of Pennsylvania | Class of 2018
Candidate for BS in Computer Science & BA in Cognitive Science
954.478.6689 | ben1sandler@gmail.com

**Comment by sandlerben** [comment]
> I guess my question is: shouldn't we continuously test the real thing - sending an actual email or downloading an actual url?

No. We should assume that when we make a call to a library function like
`smtp.Send` or `http.Get` that those library functions will magically work
correctly. We should not test external dependencies as part of _unit_ tests.

On Friday, March 4, 2016, Yoni Nachmany notifications@github.com wrote:

> Right now, determinism in mocking comes from mocking smtp functions from
> Go library, which ideally should be working, but I guess I understand that
> we should test them.
> 
> I guess my question is: shouldn't we continuously test the real thing -
> sending an actual email or downloading an actual url?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/hack4impact/audio-transcription-service/pull/24#issuecomment-192409734
> .

## 

Benjamin Sandler
University of Pennsylvania | Class of 2018
Candidate for BS in Computer Science & BA in Cognitive Science
954.478.6689 | ben1sandler@gmail.com

**Comment by yoninachmany** [comment]
> I'm actually okay with not mocking io.Copy as long as the file is deleted as part of the clean up.

To clarify: there are three Go library functions that I am using: `os.Create`, `http.Get`, and `io.Copy`

If you don't think I should mock `io.Copy`, should I still mock `os.Create` and `http.Get`?

If yes, then "clean-up" occurs in the tests, right? Thanks.

**Comment by sandlerben** [comment]
I am okay with you _only_ mocking `Get` but you might find it easier to
mock them all. If you do create a file, you may want the test to read it
and check its contents.

**Comment by yoninachmany** [comment]
Working through this issue: screen shot 2016-03-13 at 7 42 47 pm

**Comment by sandlerben** [comment]
I have no idea what that error means. What the heck is causing that?

File: transcription/transcription_test.go [hunk 5]
```diff
@@ -53,3 +55,36 @@ func TestSendEmailReturnsError(t *testing.T) {
 	err := SendEmail(username, password, host, port, to, subject, body)
 	assert.Error(err)
 }
+
+func TestDownloadFileFromURL(t *testing.T) {
```
Comment (yoninachmany) [inline transcription/transcription_test.go]: screen shot 2016-03-15 at 10 02 15 pm

Comment (yoninachmany) [inline transcription/transcription_test.go]: @sandlerben lmk what your research says. is the problem you were mentioning related to go 1.5, which the mocking library might still be running on?

Comment (sandlerben) [inline transcription/transcription_test.go]: @yoninachmany made an issue https://github.com/qur/withmock/issues/57

Let's comment out the test for now and remove the `// mock` from the http import. Please make a github issue for us to go back and fix this test. Feel free to merge once everything else is working.

Comment (yoninachmany) [inline transcription/transcription_test.go]: <3 thanks

That was the only mock test for this PR, so I will merge.

**Pull request merged**